pkg/util/hostname.go

```diff
@@ -158,9 +166,40 @@ func GetHostnameData(ctx context.Context) (HostnameData, error) {
 	hostnameErrors.Set("configuration/environment", expErr)
 
 	log.Debugf("Unable to get the hostname from the config file: %s", err)
+
+	// Try `hostname_file` config option next
+	configHostnameFilepath := config.Datadog.GetString("hostname_file")
+	if configHostnameFilepath != "" {
+		log.Debug("GetHostname trying `hostname_file` config option...")
+
+		fileHostnameProvider, found := hostname.ProviderCatalog["file"]
+		if !found {
+			return HostnameData{}, fmt.Errorf("'file' hostname provider not found when 'hostname_file' was specified")
```

Comment:
Is it desirable to return early here? I don't think it will logically cause any problems, but none of the other providers do and it makes the logic a bit misleading without carefully reading it. 

Somewhat related - but prob out of scope for this PR: This function is getting rather large and has many return points. It would be nice if we could break it up in scenario specific helpers to make it easier to follow. (it took a bit of reading to infer if this early return is going to cause problems or if other hostname sources are doing the same thing)

Reply:
> Is it desirable to return early here? I don't think it will logically cause any problems, but none of the other providers do and it makes the logic a bit misleading without carefully reading it.

I totally get it but the condition here would only happen if we can't find the built-in `'file'` provider which seems like an impossibility given that it's not a build-flag isolated provider and should be present in all build types. In theory even a `panic()`/early exit would be appropriate here as the only way this could happen would be a programming error.

> This function is getting rather large and has many return points. It would be nice if we could break it up in scenario specific helpers to make it easier to follow.

I completely agree though I don't have enough context to take on that task without some significant ramp-up and the lack of tests around this code isn't helping. Ideally, the only thing we would be doing here would be iterating over a static order of providers and returning whichever one matches (`config` would be a provider too) but I think that would be a really big undertaking fit for a card on its own.

Reply:
I think I agree with @gh123man about returning early. Even if it's an impossibility, for consistency, I'd rather have this provider behave like the others and not return early.